### PR TITLE
#161779870 Implement like and dislike functionality

### PR DESCRIPTION
**Pull request by theghostyced**
#### What does this PR do?
Implements the like and dislike functionality

#### Description of Task to be completed?
- create an article votes table
- write tests for the model
- add middleware to check if an article exists
- add endpoints
- update README.MD file
- write the functionality

#### How should this be manually tested?
Clone this branch. 
Run `npm start`
Login to the application.
Add token gotten to the headers and head over to the route `api/votes/:articleId/like`
where `:articleId` is an articleId

#### Any background context you want to provide?
The user needs to be authenticated

#### What are the relevant pivotal tracker stories?
[#161779870](https://www.pivotaltracker.com/story/show/161779870)
#### Screenshots (if appropriate)
Nil
#### Questions:
Nil

### CODE REVIEW

File: server/controllers/votes/VoteController.js [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {function} next
+   * @returns {object} Response
+   */
+  static async likeAnArticle(req, res, next) {
+    try {
+      const { articleId } = req.params;
+      const userId = req.user.id;
+
+      const dbResult = await ArticleVote.findOrCreate(
+        { where: { userId: userId, articleId: articleId } }
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: what if we used {where: {userId, articleId} } instead?

File: server/controllers/votes/VoteController.js [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+      );
+      if(!dbResult[1]) {
+        Vote.updateArticleVote(1, userId, articleId);
+        return res.status(202).json({
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: Why are we using statuscode 202?

File: server/controllers/votes/VoteController.js [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+   * @param {function} next
+   * @returns {object} Response
+   */
+  static async dislikeAnArticle(req, res, next) {
+    try {
+      const userId = req.user.id;
+      const { articleId } = req.params;
+
+      const dbResult = await ArticleVote.findOrCreate(
+        { where: { userId: userId, articleId: articleId } }
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: what are we passing as the vote object if the vote does not exist and we are trying to create it?

File: server/controllers/votes/VoteController.js [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+   * @description Handles updating a vote if user has voted
+   * @param {integer} vote
+   * @param {uuid} userId
+   * @param {uuid} articleId
+   * @returns {boolean} True or False
+   */
+  static async updateArticleVote(vote, userId, articleId) {
+    try {
+      await ArticleVote.update(
+        { vote: vote },
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js:81]: {vote}

File: server/middlewares/validations/articleValidator.js [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+   * @param {object} res
+   * @param {function} next
+   * @returns {function} next
+   */
+  static async articleExistInDatabase(req, res, next) {
+    try {
+      const { articleId } = req.params;
+      const dbResult = await Article.findOne(articleId);
+      if(dbResult) return next();
+      return res.status(404).json({
```
Comment (VictoriaOffoma) [inline server/middlewares/validations/articleValidator.js:25]: Whats the reason behind our lack of the use of the "else" statement?

File: server/models/articleVotes.js [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        type: DataTypes.DATE
+      }
+    }
+  );
+
+  ArticleVote.associate = (models) => {
+    ArticleVote.belongsTo(
+      models.Article,
+      {
+        foriegnKey: 'articleId',
```
Comment (VictoriaOffoma) [inline server/models/articleVotes.js]: it should be "foreignKey" not "foriegnKey"

File: server/controllers/votes/VoteController.js (hunk 3 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: Hi @VictoriaOffoma. Yeah, we should pass `0` back when the vote is not defined. Added this to the table models now. Will push ASAP.

File: server/routes/api/vote.js [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+
+voteRouter.post(
+  '/votes/:articleId/like',
+  JWTAuthentication.authenticateUser,
+  AricleValidator.articleExistInDatabase,
+  VoteController.likeAnArticle,
+);
+
+voteRouter.put(
+  '/votes/:articleId/unlike',
```
Comment (VictoriaOffoma) [inline server/routes/api/vote.js:19]: Difference between "Unlike" and "Dislike" is not clear on the surface.

Comment (theghostyced) [reply server/routes/api/vote.js:19]: Okay, so the unlike is used to take the vote state to its default `0` while dislike shows that you dislike an article. This situation might arise when a user like an article by mistake and wants to revert.

File: server/models/articleVotes.js (hunk 6 above)
Comment (theghostyced) [reply server/models/articleVotes.js]: Okay. Will push a fix ASAP.

File: server/controllers/votes/VoteController.js (hunk 1 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: Okay. Will push a fix for this ASAP.

File: server/middlewares/validations/articleValidator.js (hunk 5 above)
Comment (theghostyced) [reply server/middlewares/validations/articleValidator.js:25]: Okay. If `if..else` was used here, ESLint complains and says the function needs a return statement

File: server/controllers/votes/VoteController.js (hunk 2 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: This has been fixed on the latest push.

File: server/controllers/votes/VoteController.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+  }
+
+  /**
+   * @description Unlikes an article if it is already liked or disliked
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @returns {object} Response
+   */
+  static async unlikeAnAlreadyLikedOrDislikedArticle(req, res, next) {
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: This function name needs to change.

File: server/controllers/votes/VoteController.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+ */
+class Vote {
+  /**
+   * @description Hanldes Like functionality
+   * @param {object} req
+   * @param {object} res
+   * @param {function} next
+   * @returns {object} Response
+   */
+  static async reactToAnArticle(req, res, next) {
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: ReactToArticle is fine

File: server/controllers/votes/VoteController.js (hunk 8 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: Okay. What name would you prefer? @VictoriaOffoma

File: server/middlewares/validations/articleValidator.js (hunk 5 above)
Comment (VictoriaOffoma) [reply server/middlewares/validations/articleValidator.js:25]: Where is the reason?

File: server/controllers/votes/VoteController.js (hunk 9 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: Okay. Pushing the changes now.

File: server/models/articleVotes.js [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+    'ArticleVote',
+    {
+      id: {
+        type: DataTypes.UUID,
+        primaryKey: true,
+        allowNull: false,
+        defaultValue: DataTypes.UUIDV4
+      },
+      vote: {
+        type: DataTypes.INTEGER,
```
Comment (VictoriaOffoma) [inline server/models/articleVotes.js:12]: Why are we not considering ENUM instead integer?

File: server/routes/api/vote.js [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  ArticleVoteValidator.validateRequest,
+  VoteController.reactToAnArticle,
+);
+
+voteRouter.put(
+  '/votes/:articleId/unlike',
+  JWTAuthentication.authenticateUser,
+  AricleValidator.articleExistInDatabase,
+  ArticleVoteValidator.validateRequest,
+  VoteController.unlikeAnAlreadyLikedOrDislikedArticle,
```
Comment (VictoriaOffoma) [inline server/routes/api/vote.js]: can we change this name "unlikeAnAlreadyLikedOrDislikedArticle"?

File: server/controllers/votes/VoteController.js (hunk 8 above)
Comment (VictoriaOffoma) [reply server/controllers/votes/VoteController.js]: Come up with something concise.

File: server/models/articleVotes.js (hunk 10 above)
Comment (theghostyced) [reply server/models/articleVotes.js:12]: After having a few headaches trying to set a default value for the enum and not succeeding. I moved back to an integer and added a middleware that validates the `vote` input

File: server/routes/api/vote.js (hunk 11 above)
Comment (theghostyced) [reply server/routes/api/vote.js]: Do you have any name in mind? @VictoriaOffoma

File: server/controllers/votes/VoteController.js [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+      return res.status(201).json({
+        message: vote == 1 ? ARTICLE_LIKED_MSG : ARTICLE_DISLIKED_MSG
+      });
+    } catch (error) {
+      return next(error);
+    }
+  }
+
+  /**
+   * @description Unlikes an article if it is already liked or disliked
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: Modify the comment here.

File: server/helpers/responseMessages.js [hunk 13]
```diff
@@ -0,0 +1,7 @@
+/* eslint-disable max-len */
+export const ARTICLE_LIKED_MSG = 'You liked this article';
+export const ARTICLE_DISLIKED_MSG = 'You disliked this article';
+export const ARTICLE_UNLIKED_MSG = 'You unliked this article';
```
Comment (VictoriaOffoma) [inline server/helpers/responseMessages.js]: Change this name of this constant and its value.

File: server/controllers/votes/VoteController.js [hunk 14]
```diff
@@ -0,0 +1,89 @@
+import models from '../../models';
+import {
+  ARTICLE_LIKED_MSG,
+  ARTICLE_DISLIKED_MSG,
+  ARTICLE_UNLIKED_MSG
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: Change this name of this constant and its value.

File: server/controllers/votes/VoteController.js [hunk 15]
```diff
@@ -0,0 +1,89 @@
...
+    try {
+      const userId = req.user.id;
+      const { articleId } = req.params;
+      await ArticleVote.update(
+        { vote: 0 },
+        { where: { userId: userId, articleId: articleId } }
+      );
+
+      return res.status(200).json({
+        message: ARTICLE_UNLIKED_MSG
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: Change this name of this constant and its value.

File: server/controllers/votes/VoteController.js (hunk 12 above)
Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: Ok. Pushing fix now.

File: server/helpers/responseMessages.js (hunk 13 above)
Comment (theghostyced) [reply server/helpers/responseMessages.js]: Okay will push a fix ASAP

File: server/controllers/votes/VoteController.js [hunk 16]
```diff
@@ -0,0 +1,89 @@
+import models from '../../models';
+import {
+  ARTICLE_LIKE_MSG,
+  ARTICLE_DISLIKE_MSG,
+  ARTICLE_UNLIKE_MSG
```
Comment (VictoriaOffoma) [inline server/controllers/votes/VoteController.js]: what happen to my feedback about this?

Comment (theghostyced) [reply server/controllers/votes/VoteController.js]: It has been implemented.

**Pull request merged**